### PR TITLE
Documentation sprint

### PR DESCRIPTION
**Pull request by zwaldowski**
Derived from #13. Provides documentation for all members, reaching 100% coverage under [Jazzy](https://github.com/realm/jazzy).

I produced an HTML docset in the repo directory with:

```shell
~/Downloads/jazzy/bin/jazzy -m Deferred -g https://github.com/bignerdranch/Deferred -a "Big Nerd Ranch" -u https://github.com/bignerdranch --module-version=1.2.0 -r http://bignerdranch.github.io/Deferred/docs/
```

Have at it. :fire:

### CODE REVIEW

**Comment by jeremy-w** [comment]
The script in your PR message seems like a good candidate for adding to a Makefile.

File: Deferred/Deferred.swift [hunk 1]
```diff
@@ -25,9 +25,15 @@ private var DeferredDefaultQueue: dispatch_queue_t {
...
 
+/// An amount of time to wait for a deferred value.
 public enum Timeout {
+    /// Do not wait at all.
     case Now
+    /// Wait indefinitely.
     case Forever
+    /// Wait for a given number of seconds.
+    ///
+    /// :note: An `.Interval(0.0)` will achieve different results from `.Now`.
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: How do the results differ?

File: Deferred/Deferred.swift [hunk 2]
```diff
@@ -62,21 +70,38 @@ public struct Deferred<Value> {
...
+    /// * In -O builds (the default for a "Release" configuration), program
+    ///   execution will stop.
+    /// * In -Ounchecked builds, the programming error is assumed to not exist.
+    ///
+    /// If your deferred requires multiple potential fillers to race, you may
+    /// disable the precondition.
+    ///
+    /// :param: value The resolved value of the deferred.
+    /// :param: assertIfFilled If `false`, race checking is disabled.
     public func fill(value: Value, assertIfFilled: Bool = true, file: StaticString = __FILE__, line: UWord = __LINE__) {
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: The parameter `assertIfFilled` alone leaves me confused as to whether it's "assert if this `fill` call fills it" or "assert if it's already filled". It's clear from the comment and some thinking that it's the latter, but is there some alternative parameter name we could choose that would make it immediately clear at call sites, too?

File: Deferred/Deferred.swift [hunk 3]
```diff
@@ -91,7 +116,16 @@ public struct Deferred<Value> {
             }
         }
     }
-
+    
+    /**
+    Call some function once the value is determined.
+    
+    If the value is already determined, the function will be submitted to the
+    queue immediately. An `upon` call is alreadys executed asnychronously.
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: Not-a-word and typo: "alreadys executed asnychronously"

"Always"?

File: Deferred/Deferred.swift [hunk 4]
```diff
@@ -116,16 +159,43 @@ public struct Deferred<Value> {
...
     }
-
+    
+    /**
+    Waits for the value to become determined, then returns it.
+    
+    This is equivalent to unwrapping the value of calling `wait(.Forever)`, but
+    may be more efficient.
+    
+    This getter will unnecessarily block execution. It might be useful for
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: Is there a way we could apply a warning to uses of this? Would we even want to discourage its use so decisively?

File: Deferred/Deferred.swift [hunk 5]
```diff
@@ -116,16 +159,43 @@ public struct Deferred<Value> {
...
     }
 }
 
 extension Deferred {
+    /**
+    Begins another asynchronous operation with the deferred value once it
+    becomes determined.
+    
+    :param: queue A dispatch queue for starting the new operation on.
+    :param: transform A function to start a new deferred given the recieving
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: typo "recieving"

File: Deferred/Deferred.swift [hunk 6]
```diff
@@ -116,16 +159,43 @@ public struct Deferred<Value> {
...
+    value.
+    
+    :returns: A new deferred value that is determined once the recieving
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: typo "recieving"

More generally, I found this `:returns:` sentence really hard to follow. I think it might help to not talk about resolving the returned deferred. Maybe just, "The deferred value returned by `transform`."?

File: Deferred/Deferred.swift [hunk 7]
```diff
@@ -135,7 +205,15 @@ extension Deferred {
...
         return d
     }
-
+    
+    /**
+    Transforms the deferred value once it becomes determined.
+    
+    :param: queue A dispatch queue for executing the transform on.
+    :param: transform A function to create something using the deferred value.
+    :returns: A new deferred value that is determined once the recieving
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: "recieving" - I'd recommend just running a gsub across the whole project for "reciev" => "receiv" to catch all these

File: Deferred/Deferred.swift [hunk 8]
```diff
@@ -135,7 +205,15 @@ extension Deferred {
         }
         return d
     }
-
+    
+    /**
+    Transforms the deferred value once it becomes determined.
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: I think a `:note:` here and on `flatMap` to point out the distinction between them and why you'd use one vs the other might be very helpful. The param and returns docs are identical right now, which further confuses the issue.

File: Deferred/Deferred.swift [hunk 9]
```diff
@@ -146,11 +224,26 @@ extension Deferred {
 }
 
 extension Deferred {
+    /**
+    Composes the recieving value with another.
```
Comment (jeremy-w) [inline Deferred/Deferred.swift]: I just noticed that this isn't quite the same as `all([one, other])` because it returns a tuple rather than an array. :\

File: Deferred/DispatchBlockFlags.swift [hunk 10]
```diff
@@ -10,24 +10,29 @@ import Dispatch
...
     public static var allZeros: dispatch_block_flags_t {
         return dispatch_block_flags_t(0)
     }
 
+    /// Convert from a bitmask of `UInt`.
     public init(rawValue: UInt) {
         self.init(rawValue)
     }
 
+    /// Create an instance initialized with `nil`.
```
Comment (jeremy-w) [inline Deferred/DispatchBlockFlags.swift]: The docs on this differ in style from the docs in the rest of Deferred. These use an imperative ("Convert!", "Create!"), the others use a third-person indicative ("Converts", "Creates").

File: Deferred/LockProtected.swift [hunk 11]
```diff
@@ -8,25 +8,45 @@
 
 import Foundation
 
+/// A `LockProtected` holds onto a value of type T, but only allows access to it
+/// from within a locking statement. This prevents accidental unsafe access when
+/// thread safety is desired..
```
Comment (jeremy-w) [inline Deferred/LockProtected.swift]: double `.` at the end here

File: Deferred/LockProtected.swift [hunk 12]
```diff
@@ -8,25 +8,45 @@
...
     public func withReadLock<U>(block: T -> U) -> U {
         return lock.withReadLock { [unowned self] in
             return block(self.item)
         }
     }
 
+    /**
+    Give write access to the item within the given function.
+    
+    :param: block A function that writes to the contained item, and returns some
```
Comment (jeremy-w) [inline Deferred/LockProtected.swift]: Should we rename this `transform` in keeping with similar closure names in the Deferred API?

File: Deferred/DispatchBlockFlags.swift (hunk 10 above)
Comment (zwaldowski) [inline Deferred/DispatchBlockFlags.swift]: These are replicating the documentation from the stdlib. I'll change them, just pointing that out.

File: Deferred/Deferred.swift (hunk 4 above)
Comment (zwaldowski) [inline Deferred/Deferred.swift]: There might be some warnings in 2.0. All the `@availability`/`@available` that I know of cause errors.

File: Deferred/Deferred.swift (hunk 9 above)
Comment (zwaldowski) [inline Deferred/Deferred.swift]: It also, notably, allows heterogenous combination.

File: Deferred/DispatchBlockFlags.swift (hunk 10 above)
Comment (jeremy-w) [inline Deferred/DispatchBlockFlags.swift]: Maybe we should change the other docstrings to conform with the stdlib style, then?

File: Deferred/Deferred.swift (hunk 9 above)
Comment (jeremy-w) [inline Deferred/Deferred.swift]: Ah, that is nice. I'm used to Obj-C-derived types where you can always just throw them all in the `AnyObject` bucket, but that is an issue for actual Swift code.

File: Deferred/DispatchBlockFlags.swift (hunk 10 above)
Comment (zwaldowski) [inline Deferred/DispatchBlockFlags.swift]: I don't think it's necessary for all documentation to be in line with the Swift docs, we should just be consistent on our own and ideally fit some shade of BNR voice.
